doc/uhub-passwd.1
.TH UHUB-PASSWD "1" "May 2012"
.SH NAME
uhub-passwd \- small utility for manipulating passwords which are
used by uhub daemon
.SH SYNOPSIS
.B uhub-passwd
\fIfilename command \fR[...]
.SH DESCRIPTION
uhub is a high performance peer-to-peer hub for the ADC network.
Its low memory footprint allows it to handle several thousand users
on high-end servers, or a small private hub on embedded hardware.
.SH "OPTIONS"
.TP
.BI \^create
.TP
.BI \^add " username password [credentials = user]"
.TP
.BI \^del " username"
.TP
.BI \^mod " username credentials"
.TP
.BI \^pass " username password"
.TP
.BI \^list
.SH "PARAMETERS"
.TP
.B 'filename'
is a database file
.TP
.B 'username'
is a nickname (UTF\-8, up to 64 bytes)
.TP
.B 'password'
is a password (UTF\-8, up to 64 bytes)
.TP
.B 'credentials'
is one of 'admin', 'super', 'op', 'user'
.SH AUTHOR
This program was written by Jan Vidar Krey <janvidar@extatic.org>
.SH "BUG REPORTS"
If you find a bug in uhub please report it to
.B https://github.com/janvidar/uhub/issues


ChangeLog
0.5.0:
- Use TLS 1.2 and strong ciphers by default, but made this configurable.
- Fix TLS event handling which caused some busy loops
- TLS: Support certificate chains
- Fix bug #211: Better Hublist pinger support by adding the AP flag of the INF message.
- Fix bug #198:  Timers could cause infinite loops
- Sqlite3 is now mandatory
- Added mod_chat_history_sqlite and mod_chat_is_privileged.
- Support for systemd notify and journal logging
- Improved flood control counting to strictly not allow more than the given amount of messages in the configured interval.
- Optimize lookups by CID and nick.
- Added an NMDC and ADC hub redirectors written in Python.
- Fix all Clang compile warnings.
- Install uhub-passwd also.
- Add support for detecting HTTP connections to the hub. Enough to tell browsers to stop calling.
- Compile fixes for OpenBSD, including warnings about strcat.
- Fix crashing autotest due to wrong initialization of the usermanager.
- mod_topic: check argument for NULL
- rename !cleartopic to !resettopic

0.4.1:
- Converted to CMake which replaces Visual Studio project files and GNU makefiles
- Fix issues with SSL causing excessive CPU usage.
- Fix TLS/SSL handshake detection issues
- Fixed crash in mod_chat_only.
- Implemented red-black tree for better performance for certain lookups.
- Better network statistics using the !stats command.
- Improved protocol parsing, especially escape handling.
- Fix cbuffer initialization.
- Install plugins in /usr/lib/uhub, not /var/lib/uhub.
- Improved init scripts and added a upstart script.
- Work-around client security bugs by disallowing UCMD messages from being relayed.
- Added asynchronous DNS resolver.


0.4.0:
- Cleaned up code generator for config file parsing.
- Merge pull request #5 from yorhel/master
- Print error message in case of shutting down due to errors loading plugins.
- Fix Windows file read discrepancy.
- convert_to_sqlite.pl: Update to the latest SQL schema + be more Perlish
- Fix VS2010 project file - missing .c file.
- Merge https://github.com/Tilka/uhub
- use "I64u" instead of PRIu64 on Windows
- remove obsolete settings in uhub.conf
- fix uhub_itoa() and uhub_ulltoa()
- marked plugin callbacks that are not called yet
- add on_change_nick() to struct plugin_funcs
- minimal changes
- Updated init script in Debian package.
- Updated list of man pages in Debian package.
- Added man page for uhub-passwd.
- Merge branch 'master' of https://github.com/Tilka/uhub
- Fix issue with QUI messages being allowed through the hub
- don't show error on SIGTERM in select() backend
- fix dependency of 'install' target
- changed all calls to assert() to uhub_assert()
- Don't strip the U4/U6 port numbers if updated after login.
- Fix compile issue with double typedefs.
- Remove list assertion when removing element that is not in the list. Breaks autotest.
- Cleaned up command handling code, by splitting into multiple files.
- Fix bug #183 - Added proper OpenSSL license exception for the GPL.
- OMG OPTIMIZED
- use dh_prep instead of dh_clean -k
- fix double free
- use "0" instead of "false"
- fix use of uninitialized struct ip_range
- fix random crashes upon !reload
- fix command syntax
- use arg parser in !broadcast
- Fixed tiny memory leak on reload/shutdown.
- Merge https://github.com/Tilka/uhub
- fix multiple optional arguments
- small cleanup
- also clean uhub-passwd
- ignore files generated by dpkg-buildpackage
- automatically clean up plugin commands
- minimal documentation fixes
- update client software link
- update compile howto link
- fix Debian changelog
- Fix bug #158 - Added plugin for setting topic (hub description).
- Command arguments handling + cleanups


0.3.2:
- Fixed bugs in the kqueue network backend (OSX/BSD)
- Rewrote the configuration backend code.
- Added support for escaping characters in the configuration files.
- Updated the !broadcast command to send private messages instead of main chat messages.
- Adding support for redirecting clients to other hubs when they fail to log in.
- Fix some out of memory related crashes.
- Fixed minor memory leaks.


0.3.1:
- Fixed bug where !getip did not work.
- Added flood control configuration options.
- Added configuration options to disallow old/obsolete ADC clients.
- Fixed crash bugs, an freezes.
- SSL/TLS fix for tls_require configuration option.
- Fixed disconnect messages, so that clients can interpret them.
- Fixed bugs where share limits could be circumvented.
- Added support for listening to multiple ports.
- kqueue backend for Mac OS X, and BSD systems.


0.3.0:
- More user commands: ban, broadcast, mute, rules, history, myip, whoip, log
- Experimental SSL support
- Large rewrite of the network stack in order to support SSL.
- Added rule file for defining hub rules.
- Many crash fixes and other important bug fixes.
- Optimizations: O(1) timeout scheduler
- New sid allocation code.
- Added configurable server_listen_backlog (default 50).
- Added init.d scripts for RedHat/CentOS


0.2.8:
- Fix bug #13: getsockname() failure, use sockaddr from accept() instead.
- Fix bug #10: Improve logging, ensure logs are machine readable.
- Fix bug #12: asserts in adc_msg_parse -> enabled strict utf8 parsing.


0.2.7:
- Fixed a nasty crash (bug #11), Thanks Toast for finding it.
- Fix bug #9 - net_get_peer_address() failure on CentOS/Xen configurations.
- Write a log message if an operator reloads the config file.
- Don't print OK or ERROR when using '-s' or '-S' show the configuration.
- Cleanup credential string handling
- Made sure "!help" only display accessible commands.
- Cleanup in-hub command parsing
- Fix a possible crash if multiple INF messages are sent during login.
- Rewrote the mainloop to not use a timer.


0.2.6:
- Better "!uptime" command formatting.
- Better "!stats"; can display peak and current bandwidth usage.
- Added "+myip" command.
- Ensure super users and hub owners also have the operator flag set.

src/tools/uhub-passwd.c
/*
 * uhub - A tiny ADC p2p connection hub
 * Copyright (C) 2007-2014, Jan Vidar Krey
 *
 * This program is free software; you can redistribute it and/or modify
 * it under the terms of the GNU General Public License as published by
 * the Free Software Foundation; either version 3 of the License, or
 * (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU General Public License for more details.
 *
 * You should have received a copy of the GNU General Public License
 * along with this program.  If not, see <https://www.gnu.org/licenses/>.
 *
 */

#include "uhub.h"
#include "util/misc.h"
#include <sqlite3.h>

// #define DEBUG_SQL

static sqlite3* db = NULL;
static const char* command = NULL;
static const char* filename = NULL;
static const char* binary = NULL;

typedef int (*command_func_t)(size_t, const char**);

static int create(size_t argc, const char** argv);
static int list(size_t argc, const char** argv);
static int pass(size_t argc, const char** argv);
static int add(size_t argc, const char** argv);
static int del(size_t argc, const char** argv);
static int mod(size_t argc, const char** argv);

static struct commands
{
	command_func_t handle;
	const char* command;
	const char* usage;
} COMMANDS[6] = {
	{ &create, "create", "" },
	{ &list,   "list",   "" },
	{ &add,    "add",    "username password [credentials = user]" },
	{ &del,    "del",    "username" },
	{ &mod,    "mod",    "username credentials" },
	{ &pass,   "pass",   "username password" },
};

static void print_usage(const char* str)
{
	fprintf(stderr, "Usage: %s filename %s %s\n", binary, command, str);
	exit(1);
}


/**
 * Validate credentials.
 */
static const char* validate_cred(const char* cred_str)
{
	if (!strcmp(cred_str, "admin"))
		return "admin";

	if (!strcmp(cred_str, "super"))
		return "super";

	if (!strcmp(cred_str, "op"))
		return "op";

	if (!strcmp(cred_str, "user"))
		return "user";

	if (!strcmp(cred_str, "bot"))
		return "bot";

	if (!strcmp(cred_str, "ubot"))
		return "ubot";

	if (!strcmp(cred_str, "opbot"))
		return "opbot";

	if (!strcmp(cred_str, "opubot"))
		return "opubot";

	fprintf(stderr, "Invalid user credentials. Must be one of: 'bot', 'ubot', 'opbot', 'opubot', 'admin', 'super', 'op' or 'user'\n");
	exit(1);
}

static const char* validate_username(const char* username)
{
	const char* tmp;

	// verify length
	if (strlen(username) > MAX_NICK_LEN)
	{
		fprintf(stderr, "User name is too long.\n");
		exit(1);
	}

	/* Nick must not start with a space */
	if (is_white_space(username[0]))
	{
		fprintf(stderr, "User name cannot start with white space.\n");
		exit(1);
	}

	/* Check for ASCII values below 32 */
	for (tmp = username; *tmp; tmp++)
		if ((*tmp < 32) && (*tmp > 0))
		{
			fprintf(stderr, "User name contains illegal characters.\n");
			exit(1);
		}

	if (!is_valid_utf8(username))
	{
		fprintf(stderr, "User name must be utf-8 encoded.\n");
		exit(1);
	}

	return username;
}


static const char* validate_password(const char* password)
{
	// verify length
	if (strlen(password) > MAX_PASS_LEN)
	{
		fprintf(stderr, "Password is too long.\n");
		exit(1);
	}

	if (!is_valid_utf8(password))
	{
		fprintf(stderr, "Password must be utf-8 encoded.\n");
		exit(1);
	}

	return password;
}

static void open_database()
{
	int res = sqlite3_open(filename, &db);

	if (res)
	{
		fprintf(stderr, "Unable to open database: %s (result=%d)\n", filename, res);
		exit(1);
	}
}


static int create(size_t argc, const char** argv)
{
	const char* sql = "CREATE TABLE users"
		"("
			"nickname CHAR NOT NULL UNIQUE,"
			"password CHAR NOT NULL,"
			"credentials CHAR NOT NULL DEFAULT 'user',"
			"created TIMESTAMP DEFAULT (DATETIME('NOW')),"
			"activity TIMESTAMP DEFAULT (DATETIME('NOW'))"
		");";
	char* errMsg;
	int rc;

	open_database();

	rc = sqlite3_exec(db, sql, NULL, NULL, &errMsg);
	if (rc != SQLITE_OK)
	{
		fprintf(stderr, "ERROR: %s\n", errMsg);
		sqlite3_free(errMsg);
		sqlite3_close(db);
		return 1;
	}

	sqlite3_close(db);
	return 0;
}


static int sql_callback_list(void* ptr, int argc, char **argv, char **colName)
{
	int* found = (int*) ptr;
	uhub_assert(strcmp(colName[0], "nickname") == 0 && strcmp(colName[2], "credentials") == 0);
	printf("%s\t%s\n", argv[2], argv[0]);
	(*found)++;
	return 0;
}

static int list(size_t argc, const char** argv)
{
	char* errMsg;
	int found = 0;
	int rc;

	open_database();

	rc = sqlite3_exec(db, "SELECT * FROM users;", sql_callback_list, &found, &errMsg);
	if (rc != SQLITE_OK) {
#ifdef DEBUG_SQL
		fprintf(stderr, "SQL: ERROR: %s (%d)\n", errMsg, rc);
#endif
		sqlite3_free(errMsg);
		exit(1);
	}

	sqlite3_close(db);
	return 0;
}


static int add(size_t argc, const char** argv)
{
	const char* user;
	const char* pass;
	const char* cred;
	sqlite3_stmt* stmt;
	int rc;

	if (argc < 2)
		print_usage("username password [credentials = user]");

	user = validate_username(argv[0]);
	pass = validate_password(argv[1]);
	cred = validate_cred(argv[2] ? argv[2] : "user");

	open_database();

	rc = sqlite3_prepare_v2(db, "INSERT INTO users (nickname, password, credentials) VALUES(?, ?, ?);", -1, &stmt, NULL);
	if (rc != SQLITE_OK)
	{
		fprintf(stderr, "Unable to add user \"%s\": %s\n", user, sqlite3_errmsg(db));
		sqlite3_close(db);
		return 1;
	}

	sqlite3_bind_text(stmt, 1, user, -1, SQLITE_STATIC);
	sqlite3_bind_text(stmt, 2, pass, -1, SQLITE_STATIC);
	sqlite3_bind_text(stmt, 3, cred, -1, SQLITE_STATIC);

	rc = sqlite3_step(stmt);
	sqlite3_finalize(stmt);

	if (rc != SQLITE_DONE)
	{
		fprintf(stderr, "Unable to add user \"%s\"\n", user);
		sqlite3_close(db);
		return 1;
	}

	sqlite3_close(db);
	return 0;
}

static int mod(size_t argc, const char** argv)
{
	const char* user;
	const char* cred;
	sqlite3_stmt* stmt;
	int rc;

	if (argc < 2)
		print_usage("username credentials");

	user = argv[0];
	cred = validate_cred(argv[1]);

	open_database();

	rc = sqlite3_prepare_v2(db, "UPDATE users SET credentials = ? WHERE nickname = ?;", -1, &stmt, NULL);
	if (rc != SQLITE_OK)
	{
		fprintf(stderr, "Unable to set credentials for user \"%s\": %s\n", user, sqlite3_errmsg(db));
		sqlite3_close(db);
		return 1;
	}

	sqlite3_bind_text(stmt, 1, cred, -1, SQLITE_STATIC);
	sqlite3_bind_text(stmt, 2, user, -1, SQLITE_STATIC);

	rc = sqlite3_step(stmt);
	sqlite3_finalize(stmt);

	if (rc != SQLITE_DONE || sqlite3_changes(db) != 1)
	{
		fprintf(stderr, "Unable to set credentials for user \"%s\"\n", user);
		sqlite3_close(db);
		return 1;
	}

	sqlite3_close(db);
	return 0;
}

static int pass(size_t argc, const char** argv)
{
	const char* user;
	const char* password;
	sqlite3_stmt* stmt;
	int rc;

	if (argc < 2)
		print_usage("username password");

	user = argv[0];
	password = validate_password(argv[1]);

	open_database();

	rc = sqlite3_prepare_v2(db, "UPDATE users SET password = ? WHERE nickname = ?;", -1, &stmt, NULL);
	if (rc != SQLITE_OK)
	{
		fprintf(stderr, "Unable to change password for user \"%s\": %s\n", user, sqlite3_errmsg(db));
		sqlite3_close(db);
		return 1;
	}

	sqlite3_bind_text(stmt, 1, password, -1, SQLITE_STATIC);
	sqlite3_bind_text(stmt, 2, user, -1, SQLITE_STATIC);

	rc = sqlite3_step(stmt);
	sqlite3_finalize(stmt);

	if (rc != SQLITE_DONE || sqlite3_changes(db) != 1)
	{
		fprintf(stderr, "Unable to change password for user \"%s\"\n", user);
		sqlite3_close(db);
		return 1;
	}

	sqlite3_close(db);
	return 0;
}


static int del(size_t argc, const char** argv)
{
	const char* user;
	sqlite3_stmt* stmt;
	int rc;

	if (argc < 1)
		print_usage("username");

	user = argv[0];

	open_database();

	rc = sqlite3_prepare_v2(db, "DELETE FROM users WHERE nickname = ?;", -1, &stmt, NULL);
	if (rc != SQLITE_OK)
	{
		fprintf(stderr, "Unable to delete user \"%s\": %s\n", user, sqlite3_errmsg(db));
		sqlite3_close(db);
		return 1;
	}

	sqlite3_bind_text(stmt, 1, user, -1, SQLITE_STATIC);

	rc = sqlite3_step(stmt);
	sqlite3_finalize(stmt);

	if (rc != SQLITE_DONE || sqlite3_changes(db) != 1)
	{
		fprintf(stderr, "Unable to delete user \"%s\".\n", user);
		sqlite3_close(db);
		return 1;
	}

	sqlite3_close(db);
	return 0;
}

void main_usage(const char* binary)
{
	printf(
			"Usage: %s filename command [...]\n"
			"\n"
			"Command syntax:\n"
			"  create\n"
			"  add  username password [credentials = user]\n"
			"  del  username\n"
			"  mod  username credentials\n"
			"  pass username password\n"
			"  list\n"
			"\n"
			"Parameters:\n"
			"  'filename' is a database file\n"
			"  'username' is a nickname (UTF-8, up to %i bytes)\n"
			"  'password' is a password (UTF-8, up to %i bytes)\n"
			"  'credentials' is one of 'bot', 'ubot', 'opbot', 'opubot', 'admin', 'super', 'op' or 'user'\n"
			"\n"
		, binary, MAX_NICK_LEN, MAX_PASS_LEN);
}

int main(int argc, char** argv)
{
	size_t n = 0;
	binary = argv[0];
	filename = argv[1];
	command = argv[2];

	if (argc < 3)
	{
		main_usage(argv[0]);
		return 1;
	}

	for (; n < sizeof(COMMANDS) / sizeof(COMMANDS[0]); n++)
	{
		if (!strcmp(command, COMMANDS[n].command))
			return COMMANDS[n].handle(argc - 3, (const char**) &argv[3]);
	}

	// Unknown command!
	main_usage(argv[0]);
	return 1;
}
